src/lib.rs

```diff
@@ -285,29 +302,45 @@ impl<const OSC_RAW_BUF_SIZE: usize> Parser<OSC_RAW_BUF_SIZE> {
                     self.osc_raw.push(byte);
                 }
             },
-            Action::OscEnd => {
+            Action::StringEnd => {
                 let param_idx = self.osc_num_params;
                 let idx = self.osc_raw.len();
 
-                match param_idx {
-                    // Finish last parameter if not already maxed
-                    MAX_OSC_PARAMS => (),
-
-                    // First param is special - 0 to current byte index
-                    0 => {
-                        self.osc_params[param_idx] = (0, idx);
-                        self.osc_num_params += 1;
+                match self.str_start {
+                    0x1d => {
+                        match param_idx {
+                            // Finish last parameter if not already maxed
+                            MAX_OSC_PARAMS => (),
+
+                            // First param is special - 0 to current byte index
+                            0 => {
+                                self.osc_params[param_idx] = (0, idx);
+                                self.osc_num_params += 1;
+                            },
+
+                            // All other params depend on previous indexing
+                            _ => {
+                                let prev = self.osc_params[param_idx - 1];
+                                let begin = prev.1;
+                                self.osc_params[param_idx] = (begin, idx);
+                                self.osc_num_params += 1;
+                            },
+                        }
+                        self.osc_dispatch(performer, byte);
                     },
-
-                    // All other params depend on previous indexing
-                    _ => {
-                        let prev = self.osc_params[param_idx - 1];
-                        let begin = prev.1;
-                        self.osc_params[param_idx] = (begin, idx);
-                        self.osc_num_params += 1;
+                    0x18 => {
+                        self.sos_dispatch(performer);
+                    },
+                    0x1e => {
+                        self.pm_dispatch(performer);
+                    },
+                    0x1F => {
+                        self.apc_dispatch(performer);
                     },
+                    // This condition should never be hit under the
+                    // current way in which the code is ran.
+                    _ => unreachable!(),
                 }
```

Comment:
This basically feels like an embedded parser. Most of our parser is structured in a way to stream through bytes and process them as they're coming in, however rather than creating state transitions you're storing a state transition to later then match on it instead. VTE is already too slow and I'm afraid this will just amplify that problem.